src/transformers/models/blt/modular_blt.py

```diff
@@ -0,0 +1,1083 @@
+        patch_lengths = process_patch_lengths(patch_lengths, max_patch_length)
+        return concat_entropies, patch_lengths, concat_predictions
+
+    @staticmethod
+    def patch_lengths_from_entropies(
+        entropies,
+        sequence_length,
+        patch_size=None,
+        threshold=None,
+    ):
+        """
+        Computes patch lengths from token entropies.
+
+        Depending on whether a threshold is provided, the function uses either:
+        - Top-k selection based on entropy (when `threshold` is None), or
+        - Thresholding the entropy values (when `threshold` is set).
+        """
+
+        batch_size = entropies.shape[0]
+
+        # Always include token 0 and 1 as starting tokens
+        init_tokens = (
+            torch.tensor([0, 1], dtype=torch.long, device=entropies.device).unsqueeze(0).repeat(batch_size, 1)
+        )
+        offset = init_tokens.shape[1]
+
+        # Ignore first token entropy (BOS)
+        entropies = entropies[:, 1:]
+
+        if threshold is None:
+            # Use top-k entropy values to define patch start points
+            num_patches = sequence_length // patch_size
+            topk_indices = entropies.topk(num_patches - 2, dim=1).indices
+            patch_starts = topk_indices.sort(dim=1).values
+        else:
+            # Threshold the entropy values to define patch start points
+            patch_mask = entropies > threshold
+
+            seq_len = patch_mask.shape[1]
+
+            # Create patch IDs (token indices), and add a sentinel to ensure alignment
+            token_indices = torch.arange(seq_len, device=entropies.device).unsqueeze(0).expand(batch_size, -1)
+            sentinel = torch.full_like(token_indices, seq_len)
+            padded_indices = torch.cat([token_indices, sentinel], dim=1)
+
+            # Pad mask with inverse to align sentinel correctly
+            padded_mask = torch.cat([patch_mask, ~patch_mask], dim=1)
+
+            # Select indices where mask is True
+            patch_starts = padded_indices[padded_mask].reshape(batch_size, seq_len)
+            max_valid_patches = patch_mask.sum(dim=1).max()
+            patch_starts = patch_starts[:, :max_valid_patches]
+
+        # Offset patch starts to account for the two initial tokens
+        patch_start_ids = torch.cat((init_tokens, patch_starts + offset), dim=1)
+
+        # Compute patch end positions by shifting start positions
+        last_token = torch.full_like(patch_start_ids[:, :1], sequence_length - 1)
+        patch_ends = torch.cat((patch_start_ids[:, 1:] - 1, last_token), dim=1)
+
+        patch_lengths = patch_ends - patch_start_ids + 1
+
+        return patch_lengths
+
+
+class BLTForCausalLM(MllamaForCausalLM):
+    config: BLTConfig
+    supports_gradient_checkpointing = True
+    base_model_prefix = "model"
+    _tied_weights_keys = ["lm_head.weight"]
+    _no_split_modules = ["BLTTransformerLayer", "BLTLocalEncoder", "BLTLocalDecoder", "BLTGlobalTransformer"]
```

Comment:
We only need those to overwrite MLlama, the rest should be inherited from the PreTrainedModel already!
```suggestion
    config: BLTConfig
    base_model_prefix = "model"
```